File: packages/expo-dev-menu-interface/android/src/main/java/expo/interfaces/devmenu/items/DevMenuDataSource.kt
package expo.interfaces.devmenu.items

import android.os.Bundle

interface DevMenuDataSourceItem {
  fun serialize(): Bundle
}

interface DevMenuDataSourceInterface {
  val id: String

  suspend fun fetchData(): List<DevMenuDataSourceItem>
}

class DevMenuListDataSource(
  override val id: String,
  val dataFetcher: suspend () -> List<DevMenuSelectionList.Item>
) : DevMenuDataSourceInterface {
  override suspend fun fetchData(): List<DevMenuDataSourceItem> {
    return dataFetcher()
  }
}

File: packages/expo-dev-menu-interface/ios/MenuItems/DevMenuDataSource.swift
// Copyright 2015-present 650 Industries. All rights reserved.

import Foundation

@objc
public protocol DevMenuDataSourceItem {
  @objc
  func serialize() -> [String: Any]
}

public typealias DevMenuDataSourceResolver = ([DevMenuDataSourceItem]) -> Void

@objc
public protocol DevMenuDataSourceProtocol {
  var id: String { get }

  func fetchData(resolve: @escaping DevMenuDataSourceResolver)
}

@objc
public class DevMenuListDataSource: NSObject, DevMenuDataSourceProtocol {
  public var id: String
  private var dataFetcher: (@escaping ([DevMenuSelectionList.Item]) -> Void) -> Void

  public init(id: String, dataFetcher: @escaping (@escaping ([DevMenuSelectionList.Item]) -> Void) -> Void) {
    self.id = id
    self.dataFetcher = dataFetcher
  }

  public func fetchData(resolve: @escaping ([DevMenuDataSourceItem]) -> Void) {
    dataFetcher(resolve)
  }
}
